docs: modified V15-oriented shortened structure (#2)
Alzymologist/RFCs: text/0046-metadata-for-offline-signers.md

## Stakeholders

- Runtime implementors
- UI/wallet implementors
- Offline wallet implementors

All chain teams are stakeholders, as implementing this feature would require timely effort on their side and would impact compatibility with older tools.

This feature is essential for **all** offline signer tools; many regular signing tools might make use of it. In general, this RFC greatly improves security of any network implementing it, as many governing keys are used with offline signers.

Implementing this RFC would remove requirement to maintain metadata portals manually, as task of metadata verification would be effectively moved to consensus mechanism of the chain.

The idea for this RFC was brought up by runtime implementors and was extensively discussed with offline wallet implementors. It was designed in such a way that it can work easily with the existing offline wallet solutions in the Polkadot ecosystem.

## Explanation

> he FRAME metadata is providing a lot of information about a FRAME based runtime. It contains information about the pallets, the calls per pallet, the storage entries per pallet, information about runtime apis and type information about most of the types that are used in the runtime. For decoding transactions on an offline wallet we mainly require these type information. Most of the other information in the FRAME metadata are actually not required for the decoding and thus we can remove them. Thus, we have come up with a custom representation of the metadata and how this custom metadata is chunked, ensuring that we only need to send the chunks requiring for decoding a certain transaction to the offline wallet.

Reviewer: ```suggestion
The FRAME metadata provides a lot of information about a FRAME-based runtime. It contains information about the pallets, the calls per pallet, the storage entries per pallet, information about runtime APIs and type information about most of the types used in the runtime. For decoding transactions on an offline wallet, we mainly require this type of information. Most of the other information in the FRAME metadata is actually not required for the decoding, and thus, we can remove them. Thus, we have come up with a custom representation of the metadata and how this custom metadata is chunked, ensuring that we only need to send the chunks required for decoding a certain transaction to the offline wallet.
```

A reference implementation of this process is provided in [metadata-shortener](https://docs.rs/metadata-shortener/latest/metadata_shortener/) crate.

### Definitions

#### Metadata descriptor

Values for:

> 1. `u8` metadata shortening protocol version (as encoded enum variant), 

Reviewer: I'd clarify the following definiton so we make sure that is understood that we have a variant. Maybe defining the enum with `#[codec(index = 1)]` style?

Author: I'd like to keep the `V0` variant as below, because it's actually useful IMO - allows us to implement gentle crash with decent error messages.

> 2. Id of registry type encoding all extrinsic calls (u32),

Reviewer: I'd point to https://docs.rs/frame-metadata/latest/frame_metadata/v15/struct.ExtrinsicMetadata.html#structfield.call_ty

Author: ```suggestion
2. The `Id` of type of the outermost Call enum (u32) as described in https://docs.rs/frame-metadata/latest/frame_metadata/v15/struct.ExtrinsicMetadata.html#structfield.call_ty
```

3. Vector of `SignedExtension` structs as defined in Metadata V15 (vector of named (`String`) pairs of `TypeId` (`u32`))

> 4. `spec_version` `String` as found in the `RuntimeVersion` as of generating the metadata. While this information can also be found in the metadata, it is hidden in a big blob of data. To not being required to transfer this big blob of data, we directly add these information here,

Reviewer: Version is found as `u32` in metadata. I'd stick to the same type. https://github.com/paritytech/polkadot-sdk/blob/53f615de503c8936c2639f51b99019c64eab1018/substrate/primitives/version/src/lib.rs#L185

Author: This is dynamic type described in metadata; it is indeed u32 in Polkadot and in Kusama, but it could be anything, probably even enum similar to how Era is constructed - why not? - as long as it is serializable. Something like 

```
struct version {
  major: u32,
  minor: u32,
  notes: String,
}
```

would be reasonable solution.

We've really seen a couple of networks using u16 there. Not that it's all that important, but as long as our protocol would be the only limitation - it would be a limitation for our protocol; let's not enforce it unless it's enforced upstream?

The version is not exactly algebraic number - it has comparison operation, but no other algebra is defined; the primitive variable with these properties is really `String`, not integer. This is silly, we've argued a lot about this internally, but with dumb JS-style solution we seem to have all the flexibility we might want and no real downsides (as in the end from point of view of our protocol, it's just an opaque constant - key for database search).

Again, if this would be a limitation for acceptance of this RFC, it's not worth it, let's set it to whatever.

Reviewer: Let's go for `String` then, I like the rationale behind it. Could you write it there?

Author: done, check it out

}

> struct MetadataDescriptorV1 {

Reviewer: As told before, I'd rather stick to an enum and add links to docs.rs or paritytech code.

Author: I'd do it too, but unfortunately we've been asked by the reviewers to do otherwise.

}

struct MetadataDescriptorV1 {
  call_ty: u32, // TypeId
  signed_extensions: Vec<SignedExtensionsMetadata>,
  spec_version: String,
  spec_name: String,
  ss58_prefix: u16,
  decimals: u8,
  token_symbol: String
}

struct SignedExtensionMetadata {
  identifier: String,
  ty: u32, // TypeId
  additional_signed: u32, // TypeId
}
```

constitute metadata descriptor. This is minimal information that is, together with (shortened) types registry, sufficient to decode any signable transaction.

> **Note on `spec_version`**: this type is described in metadata; it is indeed u32 in Polkadot and in Kusama and most other networks and thus could be defined in typesafe manner, but theoretically it could be anything, probably even enum similar to how Era is constructed - why not? - as long as it is serializable. Something like 

Reviewer: ```suggestion
**Note on `spec_version`**: this type is described in metadata; it is indeed u32 in Polkadot and Kusama and most other networks and thus could be defined in a typesafe manner, but theoretically, it could be anything, as long as it is serializable. Something like 
```

Reviewer: Removing the example would work as well.

```
struct version {
  major: u32,
  minor: u32,
  notes: String,
}
```

would be reasonable solution that might be adopted in future.

> We've really seen a couple of networks using u16 there in the past. Not that it's all that important, but as long as our protocol would be the only limitation - it would be a limitation for our protocol; it would be unwise to enforce it here without first enforcing it upstream.
> 
> The version is not exactly algebraic number - it has comparison operation, but no other algebra is defined; the primitive variable with these properties is really `String`, not integer. This is silly, we've argued a lot about this internally, but with naive JS-style solution we have all the flexibility we might want and no real downsides (as from point of view of our protocol, it's just an opaque constant - key for database search).

Reviewer: ```suggestion
We saw a couple of networks using u16 there in the past. Not that it's all that important, but as long as our protocol would be the only limitation - it would be a limitation for our protocol; it would be unwise to enforce it here without first enforcing it upstream.

The version is not exactly an algebraic number - it has comparison operation, but no other algebra is defined; the primitive variable with these properties is really a `String`, not an integer. We've argued a lot about this internally, but with a naive JS-style solution, we have all the flexibility we might want and no real downsides (as from the point of view of our protocol, it's just an opaque constant - key for database search).
```

    u8,
    u16,
    u32,
    u64,
    u128,
    u256,
    i8,
    i16,
    i32,
    i64,
    i128,
    i256,
}

struct TypeDefCompact {
    type_param: Type,
}

struct TypeDefBitSequence {
    pub bit_store_type: Type,
    pub bit_order_type: Type,
}

```

Modularization happens thus:

1. Types registry is stripped from `docs` fields.
2. Types records are separated into chunks, with enum variants being individual chunks differing by variant index; each chunk consisting of `id` (same as in full metadata registry) and SCALE-encoded 'Type' description (reduced to 1-variant enum for enum variants). Enums with 0 variants are treated as regular types.
3. Chunks are sorted by `id` in ascending order; chunks with same `id` are sorted by enum variant index in ascending order.

```
types_registry = metadataV15.types
modularized_registry = EmptyVector<Chunk>
for (id, type) in types.registry.iterate_enumerate {
  type.doc = empty_vector
  if (type is ReduceableEnum) { // false for 0-variant enums
    for variant in type.variants.iterate {
      variant_type = Type {
        path: type.path,
        type_params: empty_vector,
        type_def: TypeDef::Variant(variants: [variant]),

>         doc: empty_vector,

Reviewer: ```suggestion
        docs: empty_vector,
```

## Testing, Security, and Privacy

> All implementations are required to follow strictly the RFC to generate the metadata hash. This includes which hash function to use and how to construct the metadata types tree. So, all implementations are following the same security criteria. As the chains will calculate the metadata hash at compile time, the build process needs to be trusted. However, this is already a solved problem in the Polkadot ecosystem by using reproducible builds. So, anyone can rebuild a chain runtime to ensure that a proposal is actually containing the changes as advertised.

Reviewer: ```suggestion
All implementations MUST follow the RFC to generate the metadata hash. This includes which hash function to use and how to construct the metadata types tree. So, all implementations are following the same security criteria. As the chains will calculate the metadata hash at compile time, the build process generally needs to be trusted. However, this is already a solved problem in the Polkadot ecosystem using reproducible builds. So, anyone can rebuild a chain runtime to ensure that a proposal is actually containing the changes as advertised.
```

Author: the second MUST - if it's assumed that actors re-build it, it's not trust, is it?

Reviewer: Hmm, should we use SHALL?

Author: idk, I just copy-pasted proposed wording from reviewers verbatim here. I don't think we need a control verb here, as we are not requiring that part. The build process is governed elsewhere and we only follow design outlined there.

Actually, I'm not sure we should talk about trusting build process here at all, but I also see that it clarifies the design a bit. Yet, if someone builds a runtime using different build system (or even uses another implementation altogether), the ecosystem would still be protected by all the other nodes of given chain requiring agreement on digest value. So what's the point of trusting build process then? We do seem to have some reliance on single build shop, but that's not really something to be proud of.
